R/utilities.R

```diff
@@ -5,6 +5,100 @@ names(rainfall_conv) = c('A>G', 'T>C', 'C>T', 'G>A', 'A>T', 'T>A', 'A>C', 'T>G',
+#' @param new_sample_sets_df Data frame of all existing and new sample sets. Required when running in default update mode. 
+#'
+#' @return
+#' @export
+#'
+#' @examples
+write_sample_set_hash = function(update=TRUE,new_sample_sets_df){
+  sample_sets_file = paste0(config::get("repo_base"),"data/metadata/level3_samples_subsets.tsv")
+  md5_file = paste0(config::get("repo_base"),"data/metadata/level3_samples_subsets_hashes.tsv")
+  if(update){
+    # load the existing file and update it using the contents of sample_sets_df as well as checking for consistency for existing sample sets
+    if(missing(new_sample_sets_df)){
+      stop("You must provide a data frame containing all the sample sets to update the digests")
+    }
+    original_digests = suppressMessages(read_tsv(md5_file))
+    
+    set_names = dplyr::select(new_sample_sets_df,-sample_id) %>% colnames()
+    #only compare for sample sets that we have in the current file
+    md5_values= c()
+    for(set_name in set_names){
+      #print(set_name)
+      this_md5 = get_samples_md5_hash(sample_set_name=set_name,sample_sets_df = new_sample_sets_df)
+      md5_values=c(md5_values,this_md5)
+      #print(this_md5)
+    }
+    all_md5 = data.frame(sample_set = set_names,new_md5_digest=md5_values)
+    oldnew = right_join(original_digests,all_md5,by="sample_set")
+    #check the rows where md5_digest is not NA (i.e. rows that were there before)
+    to_check = dplyr::filter(oldnew,!is.na(md5_digest))
+    if(any(to_check$md5_digest != to_check$new_md5_digest)){
+      problems = dplyr::filter(to_check,md5_digest!=new_md5_digest)
+      print(problems)
+      stop("some md5 digests do not match. Have these sample sets changed???")
+      
+    }
+    
+  }else{
+    # just create a file that records the md5 digests for existing sample sets
+    sample_sets = suppressMessages(read_tsv(sample_sets_file))
+    set_names = select(sample_sets,-sample_id) %>% colnames()
+    message(paste("Will log digests for",length(set_names),"sample sets"))
+    md5_values= c()
+    for(set_name in set_names){
+      print(set_name)
+      this_md5 = get_samples_md5_hash(sample_set_name=set_name)
+      md5_values=c(md5_values,this_md5)
+      print(this_md5)
+    }
+    all_md5 = data.frame(sample_set = set_names,md5_digest=md5_values)
+    write_tsv(all_md5,file=md5_file)
+  }
+
+}
+
+#' Generate an md5 hash for a set of samples to help ensure reproducibility
+#'
+#' @param these_samples_metadata Optionally provide a metadata table or any data frame with a column named sample_id that has been subset to the samples you're working with
+#' @param these_samples Optionally provide a vector of sample_id you are working with
+#' @param sample_set_name Optionall provide the name of a sample set in GAMBL and the function will load the samples from that set and provide the hash
+#' @return The md5 hash of the ordered set of sample_id
+#' @export
```

Comment:
I think this needs the `#' @import` directive to add the digest and dplyr as dependencies?